File: solutions_python/improved_library_service.py
import uuid
from datetime import datetime
from typing import List, Optional

from .book_repository import BookRepository
from .borrow_result import BorrowResult
from .fine_calculation_strategy import FineCalculationStrategy
from .loan_repository import LoanRepository
from .member_repository import MemberRepository
from .return_result import ReturnResult
from ..python_library.book import Book
from ..python_library.loan import Loan
from ..python_library.member import Member


class ImprovedLibraryService:
    """
    SOLUTION: Improved Library Service
    
    This class fixes the issues from the original code:
    1. DEBUG FIXES: Code duplication, input validation, error handling
    2. DESIGN PATTERNS: Repository pattern, Strategy pattern
    3. CODE QUALITY: Better naming, documentation, maintainability
    """
    
    # SOLUTION: Business rules configuration
    MAX_BOOKS_PER_MEMBER = 5
    LOAN_DURATION_DAYS = 14
    
    def __init__(self, book_repository: BookRepository, 
                 member_repository: MemberRepository,
                 loan_repository: LoanRepository,
                 fine_strategy: FineCalculationStrategy):
        # SOLUTION: Repository pattern - abstract data access
        self.book_repository = book_repository
        self.member_repository = member_repository
        self.loan_repository = loan_repository
        
        # SOLUTION: Strategy pattern for fine calculation
        self.fine_strategy = fine_strategy
    
    def borrow_book(self, member_id: str, book_id: str) -> BorrowResult:
        """
        SOLUTION: Improved borrow book method with better error handling and validation
        """
        # SOLUTION: Input validation - prevents null pointer exceptions
        if not member_id or not member_id.strip():
            return BorrowResult.failure("Member ID cannot be null or empty")
        if not book_id or not book_id.strip():
            return BorrowResult.failure("Book ID cannot be null or empty")
        
        try:
            # SOLUTION: Use repository pattern instead of direct list access
            member = self.member_repository.find_by_id(member_id)
            if member is None:
                return BorrowResult.failure("Member not found")
            
            book = self.book_repository.find_by_id(book_id)
            if book is None:
                return BorrowResult.failure("Book not found")
            
            # SOLUTION: Check if book is available
            if not book.is_available():
                return BorrowResult.failure("Book is not available")
            
            # SOLUTION: Business rule - check borrowing limit
            member_loans = self.loan_repository.find_by_member_id(member_id)
            if len(member_loans) >= self.MAX_BOOKS_PER_MEMBER:
                return BorrowResult.failure(f"Member has reached maximum borrowing limit of {self.MAX_BOOKS_PER_MEMBER} books")
            
            # SOLUTION: Create loan with proper validation
            loan_id = str(uuid.uuid4())
            loan = Loan(loan_id, member_id, book_id, datetime.now())
            self.loan_repository.save(loan)
            
            # SOLUTION: Update book status atomically
            book.set_available(False)
            self.book_repository.update(book)
            
            return BorrowResult.success(loan)
            
        except Exception as e:
            # SOLUTION: Specific error handling instead of generic exception catching
            return BorrowResult.failure(f"Failed to borrow book: {str(e)}")
    
    def return_book(self, loan_id: str) -> ReturnResult:
        """
        SOLUTION: Improved return book method
        """
        # SOLUTION: Input validation
        if not loan_id or not loan_id.strip():
            return ReturnResult.failure("Loan ID cannot be null or empty")
        
        try:
            # SOLUTION: Use repository pattern
            loan = self.loan_repository.find_by_id(loan_id)
            if loan is None:
                return ReturnResult.failure("Loan not found")
            
            # SOLUTION: Find and update book status
            book = self.book_repository.find_by_id(loan.get_book_id())
            if book is not None:
                book.set_available(True)
                self.book_repository.update(book)
            
            # SOLUTION: Remove loan
            self.loan_repository.delete(loan_id)
            
            return ReturnResult.success("Book returned successfully")
            
        except Exception as e:
            return ReturnResult.failure(f"Failed to return book: {str(e)}")
    
    def get_available_books(self) -> List[Book]:
        """
        SOLUTION: Improved get available books method
        """
        try:
            # SOLUTION: Use repository pattern and list comprehension for better readability
            all_books = self.book_repository.find_all()
            return [book for book in all_books if book.is_available()]
        except Exception as e:
            # SOLUTION: Return empty list instead of throwing exception
            return []
    
    def get_member_loans(self, member_id: str) -> List[Loan]:
        """
        SOLUTION: Improved get member loans method
        """
        # SOLUTION: Input validation
        if not member_id or not member_id.strip():
            return []
        
        try:
            return self.loan_repository.find_by_member_id(member_id)
        except Exception as e:
            return []
    
    def calculate_fine(self, loan_id: str) -> float:
        """
        SOLUTION: Improved fine calculation with strategy pattern
        """
        # SOLUTION: Input validation
        if not loan_id or not loan_id.strip():
            return 0.0
        
        try:
            loan = self.loan_repository.find_by_id(loan_id)
            if loan is None:
                return 0.0
            
            # SOLUTION: Use strategy pattern for fine calculation
            return self.fine_strategy.calculate_fine(loan, self.LOAN_DURATION_DAYS)
            
        except Exception as e:
            return 0.0
    
    def add_book(self, book: Book) -> bool:
        """
        SOLUTION: Add book with validation
        """
        if book is None or not book.get_id() or not book.get_id().strip():
            return False
        
        try:
            self.book_repository.save(book)
            return True
        except Exception as e:
            return False
    
    def add_member(self, member: Member) -> bool:
        """
        SOLUTION: Add member with validation
        """
        if member is None or not member.get_id() or not member.get_id().strip():
            return False
        
        try:
            self.member_repository.save(member)
            return True
        except Exception as e:
            return False

File: solutions/ImprovedLibraryService.java
package com.library.management;

import java.util.*;
import java.util.stream.Collectors;

/**
 * SOLUTION: Improved Library Service
 * 
 * This class fixes the issues from the original code:
 * 1. DEBUG FIXES: Code duplication, input validation, error handling
 * 2. DESIGN PATTERNS: Repository pattern, Strategy pattern
 * 3. CODE QUALITY: Better naming, documentation, maintainability
 */
public class ImprovedLibraryService {
    
    // SOLUTION: Repository pattern - abstract data access
    private final BookRepository bookRepository;
    private final MemberRepository memberRepository;
    private final LoanRepository loanRepository;
    
    // SOLUTION: Strategy pattern for fine calculation
    private final FineCalculationStrategy fineStrategy;
    
    // SOLUTION: Business rules configuration
    private static final int MAX_BOOKS_PER_MEMBER = 5;
    private static final int LOAN_DURATION_DAYS = 14;
    
    public ImprovedLibraryService(BookRepository bookRepository, 
                                MemberRepository memberRepository,
                                LoanRepository loanRepository,
                                FineCalculationStrategy fineStrategy) {
        this.bookRepository = bookRepository;
        this.memberRepository = memberRepository;
        this.loanRepository = loanRepository;
        this.fineStrategy = fineStrategy;
    }
    
    /**
     * SOLUTION: Improved borrow book method with better error handling and validation
     */
    public BorrowResult borrowBook(String memberId, String bookId) {
        // SOLUTION: Input validation - prevents null pointer exceptions
        if (memberId == null || memberId.trim().isEmpty()) {
            return BorrowResult.failure("Member ID cannot be null or empty");
        }
        if (bookId == null || bookId.trim().isEmpty()) {
            return BorrowResult.failure("Book ID cannot be null or empty");
        }
        
        try {
            // SOLUTION: Use repository pattern instead of direct list access
            Member member = memberRepository.findById(memberId);
            if (member == null) {
                return BorrowResult.failure("Member not found");
            }
            
            Book book = bookRepository.findById(bookId);
            if (book == null) {
                return BorrowResult.failure("Book not found");
            }
            
            // SOLUTION: Check if book is available
            if (!book.isAvailable()) {
                return BorrowResult.failure("Book is not available");
            }
            
            // SOLUTION: Business rule - check borrowing limit
            List<Loan> memberLoans = loanRepository.findByMemberId(memberId);
            if (memberLoans.size() >= MAX_BOOKS_PER_MEMBER) {
                return BorrowResult.failure("Member has reached maximum borrowing limit of " + MAX_BOOKS_PER_MEMBER + " books");
            }
            
            // SOLUTION: Create loan with proper validation
            String loanId = UUID.randomUUID().toString();
            Loan loan = new Loan(loanId, memberId, bookId, new Date());
            loanRepository.save(loan);
            
            // SOLUTION: Update book status atomically
            book.setAvailable(false);
            bookRepository.update(book);
            
            return BorrowResult.success(loan);
            
        } catch (Exception e) {
            // SOLUTION: Specific error handling instead of generic exception catching
            return BorrowResult.failure("Failed to borrow book: " + e.getMessage());
        }
    }
    
    /**
     * SOLUTION: Improved return book method
     */
    public ReturnResult returnBook(String loanId) {
        // SOLUTION: Input validation
        if (loanId == null || loanId.trim().isEmpty()) {
            return ReturnResult.failure("Loan ID cannot be null or empty");
        }
        
        try {
            // SOLUTION: Use repository pattern
            Loan loan = loanRepository.findById(loanId);
            if (loan == null) {
                return ReturnResult.failure("Loan not found");
            }
            
            // SOLUTION: Find and update book status
            Book book = bookRepository.findById(loan.getBookId());
            if (book != null) {
                book.setAvailable(true);
                bookRepository.update(book);
            }
            
            // SOLUTION: Remove loan
            loanRepository.delete(loanId);
            
            return ReturnResult.success("Book returned successfully");
            
        } catch (Exception e) {
            return ReturnResult.failure("Failed to return book: " + e.getMessage());
        }
    }
    
    /**
     * SOLUTION: Improved get available books method
     */
    public List<Book> getAvailableBooks() {
        try {
            // SOLUTION: Use repository pattern and stream API for better readability
            return bookRepository.findAll().stream()
                .filter(Book::isAvailable)
                .collect(Collectors.toList());
        } catch (Exception e) {
            // SOLUTION: Return empty list instead of throwing exception
            return new ArrayList<>();
        }
    }
    
    /**
     * SOLUTION: Improved get member loans method
     */
    public List<Loan> getMemberLoans(String memberId) {
        // SOLUTION: Input validation
        if (memberId == null || memberId.trim().isEmpty()) {
            return new ArrayList<>();
        }
        
        try {
            return loanRepository.findByMemberId(memberId);
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }
    
    /**
     * SOLUTION: Improved fine calculation with strategy pattern
     */
    public double calculateFine(String loanId) {
        // SOLUTION: Input validation
        if (loanId == null || loanId.trim().isEmpty()) {
            return 0.0;
        }
        
        try {
            Loan loan = loanRepository.findById(loanId);
            if (loan == null) {
                return 0.0;
            }
            
            // SOLUTION: Use strategy pattern for fine calculation
            return fineStrategy.calculateFine(loan, LOAN_DURATION_DAYS);
            
        } catch (Exception e) {
            return 0.0;
        }
    }
    
    /**
     * SOLUTION: Add book with validation
     */
    public boolean addBook(Book book) {
        if (book == null || book.getId() == null || book.getId().trim().isEmpty()) {
            return false;
        }
        
        try {
            bookRepository.save(book);
            return true;
        } catch (Exception e) {
            return false;
        }
    }
    
    /**
     * SOLUTION: Add member with validation
     */
    public boolean addMember(Member member) {
        if (member == null || member.getId() == null || member.getId().trim().isEmpty()) {
            return false;
        }
        
        try {
            memberRepository.save(member);
            return true;
        } catch (Exception e) {
            return false;
        }
    }
}

